script/hassfest/quality_scale_validation/has_entity_name.py

```diff
@@ -0,0 +1,180 @@
+
+from script.hassfest.model import Integration
+
+
+@dataclass
+class ClassInfo:
+    """Class info container."""
+
+    name: str
+    found: bool
+    bases: list["ClassInfo"]
+
+
+class ClassVarVisitor(ast.NodeVisitor):
+    """Visitor to report classes that do not define given classvar in their hierarchy."""
+
+    def __init__(self, var):
+        """Initialize."""
+        self._wanted_classvar = var
+        self._classes = {}
+
+    def visit_ClassDef(self, node):
+        """Visit classes."""
+        if node.name in self._classes:
+            return
+
+        def _get_class_name(node):
+            if isinstance(node, ast.Name):
+                return node.id
+
+            # generics
+            if isinstance(node, ast.Subscript):
+                return node.value.id
+
+            # enums
+            if isinstance(node, ast.Attribute):
+                return node.value.id
+
+            raise Exception("unexpected node type")  # noqa: TRY002
+
+        self._classes[node.name] = ClassInfo(name=node.name, found=False, bases=[])
+        self._classes[node.name].bases = [
+            _get_class_name(parent) for parent in node.bases
+        ]
+
+        for stmt in node.body:
+            if not isinstance(stmt, ast.Assign):
+                continue
+
+            # TODO: no unpacking, but `entity_name, foobar = True, 1 unlikely wanted
+            assert len(stmt.targets) == 1
+            var = stmt.targets[0]
+
+            if var.id == self._wanted_classvar:
+                self._classes[node.name].found = stmt.value.value
+                break
+
+    def result(self) -> list[str]:
+        """Return results."""
+
+        def _is_entity_class(info):
+            # TODO: check for platform base class names to avoid false positives?
+            return info.name.endswith("Entity") or any(
+                name for name in info.bases if name.endswith("Entity")
+            )
+
+        def _has_wanted_variable_in_hierarchy(cls_name):
+            """Return True if the class or its bases assign the wanted as True."""
+            if cls_name not in self._classes:
+                return False
+
+            return self._classes[cls_name].found or any(
+                _has_wanted_variable_in_hierarchy(name)
+                for name in self._classes[cls_name].bases
+            )
+
+        entity_classes = {
+            cls: info for cls, info in self._classes.items() if _is_entity_class(info)
+        }
+
+        # We do not care about the mro for now, return results that do not have our wanted classvar anywhere
+        return [
+            cls.name
+            for cls in entity_classes.values()
+            if not _has_wanted_variable_in_hierarchy(cls.name)
+        ]
+
+
+def validate(integration: Integration) -> list[str] | None:
+    """Validate that the integration entities have has_entity_name=True."""
+
+    wanted_variable = "_attr_has_entity_name"
+    visitor = ClassVarVisitor(wanted_variable)
```

Comment:
I feel that inside hassfest the check should not be so involved.
Such checks about inheritance would be better done inside pylint and/or as runtime checks in pytest (like translations)

My feeling is that here we should only do simple checks here, and return early (as soon as one instance is found, assume rest is OK)

```python
for file in files:
    if _has_attribute_or_argument(ast.parse(file)):
        return None
return ["Could not find use of `_attr_has_entity_name` or `has_entity_name`"]
```

```python
def _has_attribute_or_argument(module):
    for node in module.walk:
        if is_argument(node) and node.name == "has_entity_name" and node.value == True:
            return True
        if is_attribute(node) and node.name == "_attr_has_entity_name" and node.value == True:
            return True
    return False
```

Reply:
Yeah, that makes sense. I'll simplify the check, thanks for the suggestion!

Just to be sure, do you still think that we should do inheritance checks somewhere or shall we leave that out for the time being?